### PR TITLE
some hyperv related enhancement in detect-virt + vmspawn

### PR DESCRIPTION
**Pull request by poettering**
Fixes: #28001

### CODE REVIEW

File: src/basic/virt.c [hunk 1]
```diff
@@ -504,7 +504,12 @@ Virtualization detect_vm(void) {
         v = detect_vm_cpuid();
         if (v < 0)
                 return v;
-        if (v == VIRTUALIZATION_VM_OTHER)
+        if (v == VIRTUALIZATION_MICROSOFT)
+                /* QEMU sets the CPUID string to hyperv's, in case it provides hyperv enlightenments. Let's
+                 * hence not return Micrsoft here but just use the other mechanisms first to make a better
```
Comment (bluca) [inline src/basic/virt.c]: Look at your badge, it's called "Microsoft" not "Micrsoft" :-P

File: src/libsystemd/sd-id128/sd-id128.c [hunk 2]
```diff
@@ -390,3 +390,16 @@ _public_ int sd_id128_get_boot_app_specific(sd_id128_t app_id, sd_id128_t *ret)
 
         return sd_id128_get_app_specific(id, app_id, ret);
 }
+
+_public_ int sd_id128_get_invocation_app_specific(sd_id128_t app_id, sd_id128_t *ret) {
```
Comment (bluca) [inline src/libsystemd/sd-id128/sd-id128.c:394]: public API, so please add some unit tests in src/test/test-id128.c

File: src/vmspawn/vmspawn.c [hunk 3]
```diff
@@ -1294,6 +1294,24 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
                 if (strv_extend_many(&cmdline, "-uuid", SD_ID128_TO_UUID_STRING(arg_uuid)) < 0)
                         return log_oom();
 
+        /* Derive a vmgenid automatically from the invocation ID, in a deterministic way. */
+        sd_id128_t vmgenid;
+        r = sd_id128_get_invocation_app_specific(SD_ID128_MAKE(bd,84,6d,e3,e4,7d,4b,6c,a6,85,4a,87,0f,3c,a3,a0), &vmgenid);
```
Comment (bluca) [inline src/vmspawn/vmspawn.c:1299]: is that ID new? should it be documented somewhere?

Comment (poettering) [reply src/vmspawn/vmspawn.c:1299]: No, it's a private, app-specific ID, only used here. Not documented.

Comment (bluca) [reply src/vmspawn/vmspawn.c:1299]: ok

**Review by bluca — APPROVED** [review]

File: src/basic/virt.c [hunk 4]
```diff
@@ -545,7 +550,12 @@ Virtualization detect_vm(void) {
                 return v;
 
 finish:
-        if (v == VIRTUALIZATION_NONE && other)
+        /* None of the checks above gave us a clear answer, hence let's now use fallback logic: if hyperv
+         * enlightenments are available but the VMM wasn't recognized as anything yet, it's probably
+         * Microsoft. */
+        if (IN_SET(v, VIRTUALIZATION_NONE, VIRTUALIZATION_VM_OTHER) && hyperv)
```
Comment (YHNdnzj) [inline src/basic/virt.c]: Hmm, there's no need to check for `VIRTUALIZATION_VM_OTHER` here? I.e. it's unreachable as `detect_vm_zvm` would not return it, and in previous methods VIRTUALIZATION_VM_OTHER makes the search carry on.

File: man/sd_id128_get_machine.xml [hunk 5]
```diff
@@ -126,12 +132,16 @@
...
     <para>Note that <function>sd_id128_get_machine_app_specific()</function>,
-    <function>sd_id128_get_boot()</function>, <function>sd_id128_get_boot_app_specific()</function>, and
-    <function>sd_id128_get_invocation()</function> always return UUID Variant 1 Version 4 compatible IDs.
-    <function>sd_id128_get_machine()</function> will also return a UUID Variant 1 Version 4 compatible ID on
-    new installations but might not on older. It is possible to convert the machine ID non-reversibly into a
-    UUID Variant 1 Version 4 compatible one. For more information, see
+    <function>sd_id128_get_boot()</function>, <function>sd_id128_get_boot_app_specific()</function>,
+    <function>sd_id128_get_invocation()</function> and
+    <function>sd_id128_get_invocation_app_specific</function> always return UUID Variant 1 Version 4
+    compatible IDs.  <function>sd_id128_get_machine()</function> will also return a UUID Variant 1 Version 4
```
Comment (YHNdnzj) [inline man/sd_id128_get_machine.xml]:
```suggestion
    compatible IDs. <function>sd_id128_get_machine()</function> will also return a UUID Variant 1 Version 4
```

**Review by YHNdnzj — APPROVED** [review]

**Comment by YHNdnzj** [comment]
Should be mentioned in NEWS?

File: NEWS [hunk 6]
```diff
@@ -603,6 +603,12 @@ CHANGES WITH 256-rc1:
           --ssh-key-type= to optionally set up transient SSH keys to pass to the
           invoked VMs in order to be able to SSH into them once booted.
 
+        * systemd-vmspawn will no enable various "HyperV enlightenments" and
```
Comment (Vogtinator) [inline NEWS:606]: "will now"?

**Pull request merged**